hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala

```diff
@@ -227,7 +220,39 @@ case class BlockMatrixBroadcast(
 
   override def copy(newChildren: IndexedSeq[BaseIR]): BaseIR = {
     assert(newChildren.length == 1)
-    BlockMatrixBroadcast(newChildren(0).asInstanceOf[BlockMatrixIR], broadcastKind, shape, blockSize, dimsPartitioned)
+    BlockMatrixBroadcast(newChildren(0).asInstanceOf[BlockMatrixIR], inIndexExpr, outIndexExpr,
+      shape, blockSize, dimsPartitioned)
+  }
+
+  override protected[ir] def execute(hc: HailContext): BlockMatrix = {
+    val childBm = child.execute(hc)
+
+    outIndexExpr match {
+      case IndexedSeq(i, j) =>
+        val nRows = shape(0).toInt
+        val nCols = shape(1).toInt
+
+        inIndexExpr match {
+          case IndexedSeq() =>
+            val scalar = childBm.toBreezeMatrix().apply(0,0)
+            BlockMatrix.fill(hc, nRows, nCols, scalar, blockSize)
+          case IndexedSeq(`i`) => broadcastColVector(hc, childBm.toBreezeMatrix().data, nRows, nCols)
+          case IndexedSeq(`j`) => broadcastRowVector(hc, childBm.toBreezeMatrix().data, nRows, nCols)
+          case IndexedSeq(`j`, `i`) => childBm.transpose()
+        }
+    }
+  }
+
+  private def broadcastRowVector(hc: HailContext, vec: Array[Double], nRows: Int, nCols: Int): BlockMatrix = {
+    val data = ArrayBuffer[Double]()
+    (0 to nRows).foreach(_ => data ++= vec)
```

Comment:
Since we know how big `data` will be, in theory you should be able to avoid allocations from resizing `data` as you build it. I think calling `data.sizeHint(nRows * nCols)` should accomplish that.